node/impl/full/eth.go

```diff
@@ -922,62 +825,135 @@ func (a *EthModule) Web3ClientVersion(ctx context.Context) (string, error) {
 	return build.UserVersion(), nil
 }
 
-func (a *EthModule) ethCallToFilecoinMessage(ctx context.Context, tx ethtypes.EthCall) (*types.Message, error) {
-	var from address.Address
-	if tx.From == nil || *tx.From == (ethtypes.EthAddress{}) {
-		// Send from the filecoin "system" address.
-		var err error
-		from, err = (ethtypes.EthAddress{}).ToFilecoinAddress()
+func (a *EthModule) TraceBlock(ctx context.Context, blkNum string) (interface{}, error) {
+	ts, err := getTipsetByBlockNr(ctx, a.Chain, blkNum, false)
+	if err != nil {
+		return nil, err
+	}
+
+	_, trace, err := a.StateManager.ExecutionTrace(ctx, ts)
+	if err != nil {
+		return nil, xerrors.Errorf("failed to compute base state: %w", err)
+	}
+
+	tsParent, err := a.ChainAPI.ChainGetTipSetByHeight(ctx, ts.Height()+1, a.Chain.GetHeaviestTipSet().Key())
+	if err != nil {
+		return nil, fmt.Errorf("cannot get tipset at height: %v", ts.Height()+1)
+	}
+
+	msgs, err := a.ChainGetParentMessages(ctx, tsParent.Blocks()[0].Cid())
+	if err != nil {
+		return nil, err
+	}
+
+	cid, err := ts.Key().Cid()
+	if err != nil {
+		return nil, err
+	}
+
+	blkHash, err := ethtypes.EthHashFromCid(cid)
+	if err != nil {
+		return nil, err
+	}
+
+	allTraces := make([]*TraceBlock, 0, len(trace))
+	for _, ir := range trace {
+		// ignore messages from f00
+		if ir.Msg.From.String() == "f00" {
+			continue
+		}
+
+		idx := -1
+		for msgIdx, msg := range msgs {
```

Comment:
While this is correct, you _should_ be able to trust the order of the ExecutionTrace. Simply start a counter at 0, and increment it for all messages not sent by the system actor.